src/group_theory/perm/list.lean

```diff
@@ -0,0 +1,354 @@
+/-
+Copyright (c) 2021 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+
+import data.list.rotate
+import group_theory.perm.support
+
+/-!
+# Permutations from a list
+
+A list `l : list α` can be interpreted as a `equiv.perm α` where each element in the list
+is permuted to the next one, defined as `form_perm`. When we have that `nodup l`,
+we prove that `equiv.perm.support (form_perm l) = l.to_finset`, and that
+`form_perm l` is rotationally invariant, in `form_perm_rotate`.
+
+-/
+
+namespace list
+
+variables {α β : Type*}
+
+section form_perm
+
+variables [decidable_eq α] (l : list α)
+
+open equiv equiv.perm
+
+/--
+A list `l : list α` can be interpreted as a `equiv.perm α` where each element in the list
+is permuted to the next one, defined as `form_perm`. When we have that `nodup l`,
+we prove that `equiv.perm.support (form_perm l) = l.to_finset`, and that
+`form_perm l` is rotationally invariant, in `form_perm_rotate`.
+-/
+def form_perm : equiv.perm α :=
+(zip_with equiv.swap l l.tail).prod
+
+@[simp] lemma form_perm_nil : form_perm ([] : list α) = 1 := rfl
+
+@[simp] lemma form_perm_singleton (x : α) : form_perm [x] = 1 := rfl
+
+@[simp] lemma form_perm_cons_cons (x y : α) (l : list α) :
+  form_perm (x :: y :: l) = swap x y * form_perm (y :: l) :=
+prod_cons
+
+lemma form_perm_pair (x y : α) : form_perm [x, y] = swap x y := rfl
+
+lemma form_perm_apply_of_not_mem (x : α) (l : list α) (h : x ∉ l) :
+  form_perm l x = x :=
+begin
+  cases l with y l,
+  { simp },
+  induction l with z l IH generalizing x y,
+  { simp },
+  { specialize IH x z (mt (mem_cons_of_mem y) h),
+    simp only [not_or_distrib, mem_cons_iff] at h,
+    simp [IH, swap_apply_of_ne_of_ne, h] }
+end
+
+lemma form_perm_apply_mem_of_mem (x : α) (l : list α) (h : x ∈ l) :
+  form_perm l x ∈ l :=
+begin
+  cases l with y l,
+  { simpa },
+  induction l with z l IH generalizing x y,
+  { simpa using h },
+  { by_cases hx : x ∈ z :: l,
+    { rw [form_perm_cons_cons, mul_apply, swap_apply_def],
+      split_ifs;
+      simp [IH _ _ hx] },
+    { replace h : x = y := or.resolve_right h hx,
+      simp [form_perm_apply_of_not_mem _ _ hx, ←h] } }
+end
+
+@[simp] lemma form_perm_apply_last_concat (x y : α) (xs : list α) :
+  form_perm (x :: (xs ++ [y])) y = x :=
```

Comment:
Maybe a more discoverable name:
```suggestion
@[simp] lemma form_perm_cons_concat_apply_last (x y : α) (xs : list α) :
  form_perm (x :: (xs ++ [y])) y = x :=
```

Reply:
Do we also want this statement
```lean
@[simp] lemma form_perm_concat_cons_apply_last (x y : α) (xs : list α) :
  form_perm ((x :: xs) ++ [y]) y = x :=
```
which follows from the other one by rewriting with `list.concat_cons`.

Reply:
`concat_cons` is a simp lemma, so the second one simps away.